### PR TITLE
Optimize creation of sparrays from pdarrays

### PR DESCRIPTION
**Pull request by ShreyasKhandekar**
Used the algorithm provided by Engin to optimize the creation of sparse matrices from three pdarrays of rows, cols, and vals.

- [x] Testing on single and multi locale on dev cluster.

### CODE REVIEW

File: src/SparseMatrix.chpl [hunk 1]
```diff
@@ -471,32 +471,79 @@ module SparseMatrix {
...
+                        errorClass="InvalidArgumentError"
+                        );
+    if maxRow >= shape[0] || maxCol >= shape[1] then
+        throw getErrorWithContext(
+                        msg="Sparse matrix indices must be less than the shape; got (%i, %i) >= (%i, %i)".format(maxRow, maxCol, shape[0], shape[1]),
+                        lineNumber=getLineNumber(),
+                        routineName=getRoutineName(),
+                        moduleName=getModuleName(),
+                        errorClass="InvalidArgumentError"
+                        );
```
Comment (e-kayrakli) [inline src/SparseMatrix.chpl:494]: Not asking to change (b/c more precise error messages is better), but an alternative could be to throw just a single generic OOB message for both of these conditions.

File: src/SparseMatrix.chpl [hunk 2]
```diff
@@ -471,32 +471,79 @@ module SparseMatrix {
...
-                                  );
-      if rows[i] < 1 || rows[i] > shape[0] || cols[i] < 1 || cols[i] > shape[1] then
+                        msg="Sparse matrix indices must be greater than 0; got (%i, %i)".format(minRow, minCol), // TODO, change this when we start matrix from 0 instead of 1
+                        lineNumber=getLineNumber(),
+                        routineName=getRoutineName(),
+                        moduleName=getModuleName(),
+                        errorClass="InvalidArgumentError"
+                        );
+    if maxRow >= shape[0] || maxCol >= shape[1] then
+        throw getErrorWithContext(
```
Comment (e-kayrakli) [inline src/SparseMatrix.chpl:488]: indent with 2 spaces, adjust the following lines

File: src/SparseMatrix.chpl [hunk 3]
```diff
@@ -471,32 +471,79 @@ module SparseMatrix {
...
+      on loc {
+        for _srcLocId in loc.id..#numLocales {
+          const srcLocId = _srcLocId % numLocales;
+          var rowChunk = rows[rows.localSubdomain(Locales[srcLocId])];
+          var colChunk = cols[rows.localSubdomain(Locales[srcLocId])];
+          var valChunk = vals[rows.localSubdomain(Locales[srcLocId])];
+          for (r,c,v) in zip(rowChunk, colChunk, valChunk) {
+            if locDom!.parentDom.contains(r,c) {
+              if locDom!.mySparseBlock.contains(r,c) then
+                        throw getErrorWithContext(
```
Comment (e-kayrakli) [inline src/SparseMatrix.chpl:530]: indent with 2 spaces, adjust the following lines

File: src/SparseMatrix.chpl [hunk 4]
```diff
@@ -471,32 +471,79 @@ module SparseMatrix {
...
+                  );
+      SD += (r,c);
+      A[r,c] = v;
+    }
+  }
+
+
+  proc addElementsToSparseArray(ref A, ref SD, const ref rows, const ref cols, const ref vals) throws where !A.chpl_isNonDistributedArray() {
+    coforall (loc, locDom) in zip(getGrid(A),
+                                  SD._value.locDoms) {
```
Comment (e-kayrakli) [inline src/SparseMatrix.chpl:520]: I am a big fan of 80-column coding and encourage you to use that (I'll push that agenda later on :D), but this line can be just made a single line given that the function header is already overlong.

**Review by e-kayrakli — APPROVED** [review]
I have mostly cosmetic nits, otherwise looks good!

**Review by bmcdonald3 — APPROVED** [review]
Looks good, thanks Shreyas.

**Pull request merged**